vtex/faststore
# Feat: PLP - multiple templates rewrites

## What's the purpose of this pull request?

This PR aims to add multiple-page templates for friendly URLs that are added using rewrites. It also add some tests.

The documentation is in this PR:
- https://github.com/vtexdocs/dev-portal-content/pull/1157

## How it works?

#### Using Rewrites with PLP template

FastStore enables the use of [rewrites from Next.js](https://nextjs.org/docs/app/api-reference/next-config-js/rewrites) to create a custom route.
It's possible to enable the multiple-page template feature for the custom route as well by adding your rewrites function into the `faststore.config.js` file as the Next.js doc suggests:

```js
async rewrites() {
  return [
    {
      source: '/my-office',
      destination: `/office`,
    },
  ]
},
```

or

```js
async rewrites() {
  return {
    beforeFiles: [
      {
        source: '/my-office',
        destination: `/office`,
      },
    ],
    afterFiles: [],
    fallback: [],
  }
},
```

In the example above, the new `source` route `/my-office` will be rewritten to the existing `destination` route `/office`, using the `/office` template without changing the URL.
In that way, you can use the existing `destination` route `/office` to map this template in the hCMS.

**From the version of this PR, it's also possible to use the `source` route in the hCMS to map the template.**

## How to test it?

1. switch to the `vendemo` account in the cms config file (keeping the `storeframework` in the faststore.config.js just to keep the catalog).
https://github.com/vtex/faststore/blob/37bca87b4f08002bcb74b3734e9b273d8560bee2/packages/core/src/server/cms/index.ts#L45 
2. [use this page](https://vendemo.myvtex.com/admin/app/new-cms/faststore/plp/edit/57ad1d55-e094-45aa-affb-ebb3659ce518) OR create a new PLP in hCMS and use `/test/my-office`or `/office` in the input from the settings tab
<img width="1111" alt="Screenshot 2024-05-16 at 12 57 11" src="https://github.com/vtex/faststore/assets/11325562/0dd2b739-58be-429f-8bdc-9eaae77b7a4a">
3. publish your page and run the project locally with `yarn dev`
4. both URLs (`/test/my-office`or `/office`) should show the new template.


## References

PLP doc
- https://github.com/vtexdocs/dev-portal-content/pull/1157
